### PR TITLE
Use multistage builds in a Dockerfile to compile and create an image from source

### PR DESCRIPTION
**Pull request by jsternberg**
The multistage build lets us build the image inside of one container and
then copy over the results to a new image so the build dependencies
aren't needed in the final container.

The resulting container should be identical to the official built ones
for every release, but coming from source instead.

- [x] Rebased/mergable
- [x] Tests pass

### CODE REVIEW

File: docker-init-influxdb.sh [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+AUTH_ENABLED="$INFLUXDB_HTTP_AUTH_ENABLED"
+
+if [ -z "$AUTH_ENABLED" ]; then
+	AUTH_ENABLED="$(grep -iE '^\s*auth-enabled\s*=\s*true' /etc/influxdb/influxdb.conf | grep -io 'true' | cat)"
+else
+	AUTH_ENABLED="$(echo ""$INFLUXDB_HTTP_AUTH_ENABLED"" | grep -io 'true' | cat)"
+fi
+
+INIT_USERS=$([ ! -z "$AUTH_ENABLED" ] && [ ! -z "$INFLUXDB_ADMIN_USER" ] && echo 1 || echo)
```
Comment (mark-rushakoff) [inline docker-init-influxdb.sh]: Is there a reason to use `! -z` instead of `-n`?

And are the environment variables in this file documented anywhere?

Comment (jsternberg) [reply docker-init-influxdb.sh]: Yea, they're documented in the docker docs themselves. I copied the file directly from there.

I'm not sure at the moment how we should keep the two connected with each other. I just thought this feature was cool and a decent way to allow someone to more easily build a Docker image from master.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+COPY Godeps /go/src/github.com/influxdata/influxdb/Godeps
+WORKDIR /go/src/github.com/influxdata/influxdb
+RUN gdm restore
+COPY . /go/src/github.com/influxdata/influxdb
+RUN go install ./cmd/...
+
+FROM debian:stretch
+COPY --from=builder /go/bin/* /usr/bin/
+RUN mkdir -p /etc/influxdb && \
+    influxd config > /etc/influxdb/influxdb.conf
```
Comment (jwilder) [inline Dockerfile]: This shouldn't be `influxd config`, should copy in the the `/etc/config.sample.toml` as the config.

Comment (jsternberg) [reply Dockerfile]: Done.

**Comment by jsternberg** [comment]
@jwilder I updated this to use go 1.9.2 and moved the docker specific files to a folder instead.

**Pull request merged**